### ISSUE TITLE
Immutables integration not working correctly with custom Value.Style typeImmutable

### ISSUE THREAD

**Issue by ThLeu** [mapstruct/mapstruct#1924]
- [x] Is this an issue (and hence not a question)? 

Hi, This issue basically concerns what was already mentioned in this comment (https://github.com/mapstruct/mapstruct/issues/782#issuecomment-416137293) and as I didn't find an issue yet, here it is now because we ran into this problem aswell.

So the issue is concerning this exact line here:
https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/spi/ImmutablesBuilderProvider.java#L83

The Immutables integration only works correctly if Immutables generated classes are configured to use the default "Immutables" prefix (which is hardcoded on this line).

This caused us some severe issues because MapStruct will in some cases not generate an implementation and also not report any errors or anything (even with MapStruct policies set to ERROR), the build was even successful as only the implementation was missing. This happened when we used an Immutable class inside another Immutable class (both not with the default prefix style), but I am not sure if the problem only occurs with nested Immutables classes.

We also defined a custom annotation which uses the Immutables `@Value.Style` annotation as a meta-annotation.

```
@Value.Style(
      // The generated Immutable will have 'Immutable' suffix instead of prefix
      typeImmutable = "*Immutable"
public @interface CustomStyleAnnotation {
}
```

So in short the bug is that that `@Value.Style` `typeImmutable`value is not considered when working with Immutables.

The addition that it can also be a meta-annotation in a custom annotation is probably just a detail but it also be considered.

Best regards

**Comment by sjaakd** [comment]
@ThLeu .. Thanks for your comment.

The idea behind the builder provider is that you can write your own ones to cope with these kind of customisation issues. We can't possibly cater for every permutation of a framework.  

Does our ref guide offer enough information for you to use our SPI and write your own provider?

**Comment by ThLeu** [comment]
Thank you sjaakd for your answer.

Ok I understand that using our own implementation of the SPI would be a solution and we might consider that, thanks for that hint.
But as MapStruct is already providing a default implementation for the Immutables library with the ImmutablesBuilderProvider I think it would make sense to extend that to also consider the default way that Immutables generated classes can be customised. The only customisation is to my knowledge the `@Value.Style(typeImmutable)` so the current implementation of the ImmutablesBuilderProvider would probably only need to be extended by a few lines.
If I find the time at some point I might provide a PR for that (but please don't count on that if you want to do it). Would you be ok with implementing it if someone provided the PR for it?

**Comment by sjaakd** [comment]
> But as MapStruct is already providing a default implementation for the Immutables library with the ImmutablesBuilderProvider I think it would make sense to extend that to also consider the default way that Immutables generated classes can be customised. The only customisation is to my knowledge the `@Value.Style(typeImmutable)` so the current implementation of the ImmutablesBuilderProvider would probably only need to be extended by a few lines.

I'm curious how that would look like. If I understand you correctly, `@Value.Style(typeImmutable)` is a configuration option. SPI does not allow passing any parameters to reflect upon such configuration properties. I guess the current SPI solution then needs to deal with this option. Perhaps you can derive it from the `@Value.Style` annotation, but otherwise it will be tricky.

**Comment by filiphr** [comment]
@sjaakd `@Value.Style` is an annotation on the class itself. So as @ThLeu proposed we could try to get it in the `ImmutablesBuilderProvider`.  

@ThLeu I would say that if there is a PR for that logic we would merge it in our codebase. The reason why we have a fix `Immutable` in the `ImmutablesBuilderProvider` was that none of use knows the Immutables options well and we were relying on the community to provide us more guidance via issues.

**Comment by Chessray** [comment]
I've come here due to a related issue:

On my new project, we want to combine MapStruct and Immutables. In proper DRY fashion, we've introduced our own custom immutable annotation using the mechanism described [here](http://immutables.github.io/style.html#custom-immutable-annotation). However, the types annotated in this fashion can't be used as target type for a mapping method, because MapStruct doesn't realise that this is actually an Immutables-related annotation and claims that there is no setter on the return type. The only workaround we've found so far is to use the class generated by the Immutables processor as return type instead of the annotated one.

I'm leavig this comment here, because I'm not sure whether this observation is the same issue or constitutes a new one. I'd really appreciate your opinions on this @filiphr @sjaakd.

**Comment by sjaakd** [comment]
@Chessray don't have a computer available at the moment. Will look in more detail  into this as soon as I've the opportunity. But may take a while. Did you (by any chance) look into the accessor naming strategy SPI ?

**Comment by filiphr** [comment]
@Chessray your issue is different. From what I understood you are using immutables as meta annotations. However, we don't do that check. 

We currently do:

https://github.com/mapstruct/mapstruct/blob/2e2c20fed7502c2cb86d20ede130dfa499813a48/processor/src/main/java/org/mapstruct/ap/spi/ImmutablesBuilderProvider.java#L52-L67

To find the immutables builder. You can provide your own `ImmutablesBuilderProvider` that will lookup the annotation from the meta annotations as well. You can also provide a PR that would do that in our current `ImmutablesBuilderProvider`

**Comment by Drevsh** [comment]
Hey I created a pull request for this issue: https://github.com/mapstruct/mapstruct/pull/3189

Would love some feedback. Prefix/Postfix/Infix and no `@Value.Style` annotation are correctly handled. I know that currently `@Value.Style( typeAbstract = ...)` is still not handled, but it felt out of scope. Maybe I'll create a new issue for that.

**Comment by filiphr** [comment]
Thanks @Drevsh  and @sleicht for your work on this. I've added it to the 1.6 milestone and I'll try to review the PR soon
